### PR TITLE
Messaging feature

### PR DESCRIPTION
**Pull request by Mari-Hayashi**
Implemented messaging feature.

* Users can send private messages to each other.
* The chat area is reloaded automatically when a user receives a new message.

This PR does not include:
* Highlighting unread messages.
* Pagination/Scroll view of the chat area.

### CODE REVIEW

File: src/client/src/Domains/Messages/ConversationPreview/index.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        },
+        Selected: {
+            backgroundColor: '#efefef',
+            borderColor: theme.palette.primary.main,
+        },
+    })
+);
+
+// Convert dateString to a readable formatting.
+const getDateString = (dateString: string): string => {
```
Comment (Mari-Hayashi) [inline src/client/src/Domains/Messages/ConversationPreview/index.tsx]: I'm using exactly the same function in MessageDialog. Is there any way to write this function in one place and call from both ConversationPreview/index.tsx and MessageDialog/index.tsx?

File: src/client/src/Domains/Messages/api/index.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+export async function getConversationList() {
+    return API.get<{ conversationList: IConversation[] }>('message/getConversationList');
+}
+
+export async function createMessage(newMessageForm: INewMessageForm) {
+    return API.get < { user: IUser, error: string } >(`user/get-by-email/${newMessageForm.email}`);
+}
+
+const serverPath = 'http://localhost:5000/';
```
Comment (Mari-Hayashi) [inline src/client/src/Domains/Messages/api/index.tsx]: Where should we get serverPath? Originally I get from src/api.tsx, but it's not working after I merged with main.

File: src/server/src/server/index.ts [hunk 3]
```diff
@@ -10,6 +12,22 @@ import logger from '@shared/Logger';
 
 // Start the server
 const port = Number(process.env.PORT || 5000);
-app.listen(port, () => {
+const server = app.listen(port, () => {
     logger.info('Server started on port: ' + port);
 });
+
+const clientPath = 'http://localhost:3000';
```
Comment (Mari-Hayashi) [inline src/server/src/server/index.ts]: Where should we get the client path? Originally I get from src/app, but it's not working after I merged with main.

File: src/client/src/Domains/Messages/ConversationPreview/index.tsx (hunk 1 above)
Comment (johan1505) [reply src/client/src/Domains/Messages/ConversationPreview/index.tsx]: In the project, there is a package called date-fns (https://date-fns.org/v2.16.1/docs/Getting-Started). Read their documentation, I think they might have a function that does this. If you find such function, then use it. You can see an example of me using this package in client/src/utils/format.ts. If they don't have such a function then add your function to the format.ts file previously mentioned and export from there.

File: src/client/src/Domains/Messages/ConversationList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+import useApi from 'hooks/useApi';
+import { AuthContext } from 'Contexts/AuthContext';
+import Loader from 'Components/Loader';
+import { IUser } from 'Domains/Accounts/api';
+import { IConversation, IMessage, getConversationList, io } from 'Domains/Messages/api';
+import NewMessageForm from 'Domains/Messages/NewMessageForm';
+import ConversationPreview from 'Domains/Messages/ConversationPreview';
+
+interface props {
```
Comment (johan1505) [inline src/client/src/Domains/Messages/ConversationList/index.tsx]: Rename to ConversationListProps

File: src/client/src/Domains/Messages/ConversationPreview/index.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+import Card from '@material-ui/core/Card';
+import Grid from '@material-ui/core/Grid';
+import Typography from '@material-ui/core/Typography';
+import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
+
+import { IUser } from 'Domains/Accounts/api';
+import { IConversation } from 'Domains/Messages/api';
+
+interface Props {
```
Comment (johan1505) [inline src/client/src/Domains/Messages/ConversationPreview/index.tsx]: Rename to ConversationPreviewProps

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import Typography from '@material-ui/core/Typography';
+
+import useApi from 'hooks/useApi';
+import { AuthContext } from 'Contexts/AuthContext';
+import Loader from 'Components/Loader';
+import { IUser } from 'Domains/Accounts/api';
+import MessageDialog from 'Domains/Messages/MessageDialog';
+import { IMessage, getMessages, io } from 'Domains/Messages/api';
+
+interface props {
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: MessagesProps

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        if (receiver) sendRequest();
+    }, [receiver, sendRequest]);
+
+    // Reload messaging component if new message is created between the
+    // logged-in user and the current receiver.
+    io.on('chat', (message: IMessage) => {
+        if (receiver &&
+            ((message.sender.id === receiver.id && message.receiver.id === user?.userId)
+                || (message.receiver.id === receiver.id && message.sender.id === user?.userId))) {
+            sendRequest();
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: Curious, is it completely necessary to send the request? Can't we append the new message to the current list of messages? To me it looks kind of weird, having the page showing the loading spinner every time either person sends a message.

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                    {isLoading ? (
+                        <Loader />
+                    ) : (
+                            <div>
+                                <Typography variant='h6' color='primary'>
+                                    Message with {receiver.firstName} {receiver.lastName}
+                                    </Typography>
+                                    <Grid>
+                                        {/* Render all messages. */}
+                                        {messages.map((message, key) => (
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: rename key to index

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    const [messages, setMessages] = React.useState<IMessage[]>([]);
+
+    const { user } = React.useContext(AuthContext);
+
+    const request = React.useCallback(() => getMessages(receiver), [receiver]);
+    const [sendRequest, isLoading] = useApi(request, {
+        onSuccess: (response) => {
+            setMessages(response.data.messages);
+        },
+        onFailure: (error, response) => {
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: Since you are only outputting the error, you can remove this onFailure function. The default onFailure included in useApi will take care of outputting the error.

File: src/client/src/Domains/Messages/MessageDialog/index.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                <Grid container spacing={4} alignItems='center' justify='center'>
+                    <Grid
+                        item
+                        container
+                        direction='column'
+                        spacing={1}
+                        md={9}
+                        xs={12}
+                    >
+                        <Grid item>
```
Comment (johan1505) [inline src/client/src/Domains/Messages/MessageDialog/index.tsx]: is this grid really needed?

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+const formSchema = yup.object({
+    message: yup.string().required('Message cannot be empty.'),
+});
+
+const messageInitialValues: IMessageSendForm = {
+    message: ''
+};
+
+interface props {
```
Comment (johan1505) [inline src/client/src/Domains/Messages/MessageSendForm/index.tsx]: MessageSendFormProps

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    const [message, setMessage] = React.useState<IMessageSendForm>(messageInitialValues);
+
+    const { user } = React.useContext(AuthContext);
+
+    const request = React.useCallback(() => sendMessage(message, receiver), [message, receiver]);
+    const [sendRequest, isLoading] = useApi(request, {
+        onSuccess: () => {
+            io.emit('chat', { message: message, receiver: receiver, sender: {id: user?.userId } });
+        },
+        onFailure: (error, response) => {
```
Comment (johan1505) [inline src/client/src/Domains/Messages/MessageSendForm/index.tsx]: Don't think you need this

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                                <Field
+                                    name='message'
+                                    label='Message'
+                                    component={TextFormField}
+                                />
+                            </Grid>
+                            <Grid item md={3} xs={2}>
+                                <SubmitButton fullWidth={true} type='submit' isLoading={isLoading} disabled={receiver === undefined}>
+                                    Send
+                                    {isLoading && <Loader size={20} />}
```
Comment (johan1505) [inline src/client/src/Domains/Messages/MessageSendForm/index.tsx]: No need to add this logic or Loader. It's taken care in SubmitButton now

File: src/client/src/Domains/Messages/NewMessageForm/index.tsx [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+                                    <Field
+                                        name='email'
+                                        label='Email'
+                                        component={TextFormField}
+                                    />
+                                </Grid>
+                            </Grid>
+                            <Grid container item xs={12}>
+                                <SubmitButton type='submit' isLoading={isLoading}>
+                                    Submit
```
Comment (johan1505) [inline src/client/src/Domains/Messages/NewMessageForm/index.tsx]: Remove Submit label. The default label is always submitted inside of SubmitButton

File: src/client/src/Domains/Messages/NewMessageForm/index.tsx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+                                    {isLoading && <Loader size={20} />}
```
Comment (johan1505) [inline src/client/src/Domains/Messages/NewMessageForm/index.tsx]: Remove this

File: src/client/src/Domains/Messages/api/index.tsx (hunk 2 above)
Comment (johan1505) [reply src/client/src/Domains/Messages/api/index.tsx]: I think you need to add it to that file and export it?

File: src/server/src/modules/message.ts [hunk 16]
```diff
@@ -0,0 +1,203 @@
+import { Message } from '@entities/message';
+import { User } from '@entities/user';
+import { getRepository } from 'typeorm';
+
+/**
+ * @description Return a boolean value checking if the user with the given id exists.
+ * @param {number} userId - id of the user.
+ * @returns Promise
+ */
+const userIdExists = async (userId: number) => {
```
Comment (johan1505) [inline src/server/src/modules/message.ts]: I think this function should go in the modules/user.ts

File: src/server/src/modules/message.ts [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+        .getMany();
+
+    getMessagesResult.errorMessage = "Successful";
+    getMessagesResult.result = messages;
+
+    return getMessagesResult;
+};
+
+/**
+ * @description Get an array of User who has communicated with the user with the given id
```
Comment (johan1505) [inline src/server/src/modules/message.ts]: array of Users who have communicated..

File: src/server/src/routes/user/index.ts [hunk 18]
```diff
@@ -19,6 +19,12 @@ interface ISignUpRequest extends Request {
     };
 }
 
+interface IGetUserByEmailRequest extends Request {
```
Comment (johan1505) [inline src/server/src/routes/user/index.ts]: I don't think you need this. You are getting the email from the params

File: src/server/src/routes/message/index.ts [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+interface sendMessageRequest extends Request {
+    body: {
+        message: string;
+        receiverId: number;
+    };
+}
+
+interface getMessageRequest extends Request {
+    body: {
+        communicatorId: number;
```
Comment (johan1505) [inline src/server/src/routes/message/index.ts]: I don't think this property is ever used so the interface is irrelevant

File: src/server/src/server/index.ts (hunk 3 above)
Comment (johan1505) [reply src/server/src/server/index.ts]: Add the export keyword to a src/server/src/server/index.ts on line 19. You get it from there then

**Review by johan1505 — CHANGES_REQUESTED** [review]
Left comments. Next time to the pr into the 'dev' branch. Also, in src run `docker-compose -f docker-compose.test.yml up` to run tests. Make sure they are passing. It would be great if you could write tests for your backend API (you can look at server/src/routes/Sample/sample-routes.test.ts to see how to write tests).

**Comment by Mari-Hayashi** [comment]
> Left comments. Next time to the pr into the 'dev' branch. Also, in src run `docker-compose -f docker-compose.test.yml up` to run tests. Make sure they are passing. It would be great if you could write tests for your backend API (you can look at server/src/routes/Sample/sample-routes.test.ts to see how to write tests).

Thank you! Do you mind if I add test cases in the next PR? This PR is already so large.

File: src/client/src/Domains/Messages/ConversationPreview/index.tsx (hunk 1 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/ConversationPreview/index.tsx]: Done.

File: src/client/src/Domains/Messages/api/index.tsx (hunk 2 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/api/index.tsx]: Done.

File: src/server/src/server/index.ts (hunk 3 above)
Comment (Mari-Hayashi) [reply src/server/src/server/index.ts]: ?? This file is src/server/src/server/index.ts. Should I just keep this here?

File: src/client/src/Domains/Messages/ConversationList/index.tsx (hunk 4 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/ConversationList/index.tsx]: Done.

File: src/client/src/Domains/Messages/ConversationPreview/index.tsx (hunk 5 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/ConversationPreview/index.tsx]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 6 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 7 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: When I just append the message to the "messages" list, the html component (At the bottom of this file) is not updated.

File: src/server/src/modules/message.ts (hunk 16 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/message.ts]: Done.

File: src/server/src/modules/message.ts (hunk 17 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/message.ts]: Done.

File: src/server/src/routes/user/index.ts (hunk 18 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/user/index.ts]: Done.

File: src/server/src/routes/message/index.ts (hunk 19 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/message/index.ts]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 8 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 9 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Done.

File: src/client/src/Domains/Messages/MessageDialog/index.tsx (hunk 10 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/MessageDialog/index.tsx]: Done.

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx (hunk 11 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/MessageSendForm/index.tsx]: Done.

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx (hunk 12 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/MessageSendForm/index.tsx]: Done.

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx (hunk 13 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/MessageSendForm/index.tsx]: Done.

File: src/client/src/Domains/Messages/NewMessageForm/index.tsx (hunk 14 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/NewMessageForm/index.tsx]: Done.

File: src/client/src/Domains/Messages/NewMessageForm/index.tsx (hunk 15 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/NewMessageForm/index.tsx]: Done.

File: src/server/src/server/index.ts (hunk 3 above)
Comment (johan1505) [reply src/server/src/server/index.ts]: oh oopse sorry. I meant can you move the clientPath variable to src/app/index.ts and then export from there. Also add in in that file's line 24.  `app.use(cors({ credentials: true, origin: clientPath' }));`

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 20]
```diff
@@ -10,23 +10,20 @@ import { IUser } from 'Domains/Accounts/api';
...
 
-interface props {
+interface MessagesProps {
     receiver: IUser | undefined;
 }
 
-function Messages({ receiver }: props) {
-    const [messages, setMessages] = React.useState<IMessage[]>([]);
+function Messages({ receiver }: MessagesProps) {
+    var [messages, setMessages] = React.useState<IMessage[]>([]);
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: I don't think this should be a var?

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 7 above)
Comment (johan1505) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: uh that's weird. are you using the setMessages function to append the new message? If so, it should work. Like this:

`setMessages ((prevMessages) => [...prevMessages, message]);
`

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 21]
```diff
@@ -40,6 +37,7 @@ function Messages({ receiver }: props) {
             ((message.sender.id === receiver.id && message.receiver.id === user?.userId)
                 || (message.receiver.id === receiver.id && message.sender.id === user?.userId))) {
             sendRequest();
+            // messages.push(message);
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: oh yah that doesn't work. you gotta use the setMessages function

**Review by johan1505 — CHANGES_REQUESTED** [review]
left more comments

**Comment by johan1505** [comment]
> > Left comments. Next time to the pr into the 'dev' branch. Also, in src run `docker-compose -f docker-compose.test.yml up` to run tests. Make sure they are passing. It would be great if you could write tests for your backend API (you can look at server/src/routes/Sample/sample-routes.test.ts to see how to write tests).
> 
> Thank you! Do you mind if I add test cases in the next PR? This PR is already so large.

Sure, let's do that.

File: src/server/src/server/index.ts (hunk 3 above)
Comment (Mari-Hayashi) [reply src/server/src/server/index.ts]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 7 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 20 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Done.

File: src/client/src/Domains/Messages/GetMessages/index.tsx [hunk 22]
```diff
@@ -30,16 +30,21 @@ function Messages({ receiver }: MessagesProps) {
...
-                || (message.receiver.id === receiver.id && message.sender.id === user?.userId))) {
-            sendRequest();
-            // messages.push(message);
-        }
-    });
+    React.useEffect(() => {
+        // Update io listener by every receiver update.
+        io.removeListener('message_area');
+        io.on('message_area', (message: IMessage) => {
+            // Reload messaging component if new message is created between the
```
Comment (johan1505) [inline src/client/src/Domains/Messages/GetMessages/index.tsx]: wait, so it keeps doing a reload even though the request is not being sent?

Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: What do you mean?

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx [hunk 23]
```diff
@@ -34,8 +34,8 @@ function MessageSendForm({ receiver }: MessageSendFormProps) {
     const request = React.useCallback(() => sendMessage(message, receiver), [message, receiver]);
     const [sendRequest, isLoading] = useApi(request, {
         onSuccess: () => {
-            io.emit('chat', {
-                message: message,
+            io.emit('new_message', {
+                message: message.message,
```
Comment (johan1505) [inline src/client/src/Domains/Messages/MessageSendForm/index.tsx]: Maybe a better name for the message property of the message object would be content. so message.content

File: src/client/src/Domains/Messages/GetMessages/index.tsx (hunk 22 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: It reloads when a new message is created between the logged-in user and the current receiver.

Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Or, it appends a new message to the "messages" list if the new message is sent by the logged-in user or the currently selected receiver.

Comment (johan1505) [reply src/client/src/Domains/Messages/GetMessages/index.tsx:37]: oh so, the feature doesn't look like let's say FB messenger chat so far. right? Users have to fill out the NewMessageForm to create a message

Comment (johan1505) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: oh okay so it's not reloading the component then, it's just appending the message to the list of messages. If so, maybe rephrase the comment on line 37

Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/GetMessages/index.tsx]: Done.

File: src/client/src/Domains/Messages/MessageSendForm/index.tsx (hunk 23 above)
Comment (Mari-Hayashi) [reply src/client/src/Domains/Messages/MessageSendForm/index.tsx]: Done.

**Review by johan1505 — APPROVED** [review]
